File: casino.py
"""Cassino: the Hungarian two-player version, 52-card French deck.

Cards are strings: rank, then suit. Ranks: A 2 3 4 5 6 7 8 9 10 J Q K.
Suits: S H D C.
"""
from dataclasses import dataclass
from itertools import combinations
from typing import NamedTuple, Optional

_RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]
_VALUES = {rank: i + 1 for i, rank in enumerate(_RANKS)}


def value(card):
    """A card's value: A=1, 2..10 face value, J=11, Q=12, K=13."""
    return _VALUES[card[:-1]]


class Move(NamedTuple):
    hand: frozenset
    table: frozenset


@dataclass(frozen=True)
class State:
    hands: tuple
    table: tuple
    talon: tuple
    piles: tuple
    sweeps: tuple
    player: int
    last_capturer: Optional[int] = None


def new_deal(deck, first=0):
    """Deal from `deck`: 3 cards to `first`, 3 to the other, 4 to the table."""
    d = tuple(deck)
    hands = [None, None]
    hands[first] = d[0:3]
    hands[1 - first] = d[3:6]
    return State(
        hands=tuple(hands),
        table=d[6:10],
        talon=d[10:],
        piles=((), ()),
        sweeps=(0, 0),
        player=first,
        last_capturer=None,
    )


def _nonempty_subsets(cards):
    cards = list(cards)
    for size in range(1, len(cards) + 1):
        yield from combinations(cards, size)


def legal_moves(state):
    """Every legal Move for the player to move."""
    hand = state.hands[state.player]
    table = state.table

    moves = {Move(frozenset((c,)), frozenset()) for c in hand}

    table_subsets_by_sum = {}
    for t in _nonempty_subsets(table):
        s = sum(value(c) for c in t)
        table_subsets_by_sum.setdefault(s, []).append(t)

    for h in _nonempty_subsets(hand):
        s = sum(value(c) for c in h)
        for t in table_subsets_by_sum.get(s, ()):
            moves.add(Move(frozenset(h), frozenset(t)))

    return moves


def play(state, move):
    """The state after `move`, including whatever the rules make happen next."""
    if move not in legal_moves(state):
        raise ValueError(f"illegal move: {move}")

    player = state.player
    other = 1 - player

    hands = list(state.hands)
    hands[player] = tuple(c for c in hands[player] if c not in move.hand)

    table = state.table
    table_was_empty = len(table) == 0
    piles = list(state.piles)
    sweeps = list(state.sweeps)
    last_capturer = state.last_capturer

    captured = bool(move.table)
    if captured:
        table = tuple(c for c in table if c not in move.table)
        piles[player] = piles[player] + tuple(move.hand | move.table)
        last_capturer = player
        if not table:
            sweeps[player] += 1
    else:
        (placed,) = move.hand
        table = table + (placed,)

    next_player = (
        player if (not captured and table_was_empty and hands[player]) else other
    )

    talon = state.talon
    if not hands[0] and not hands[1]:
        if talon:
            lead = last_capturer if last_capturer is not None else next_player
            if lead == 0:
                hands[0], hands[1] = talon[0:3], talon[3:6]
            else:
                hands[1], hands[0] = talon[0:3], talon[3:6]
            talon = talon[6:]
            next_player = lead
        elif table:
            collector = last_capturer if last_capturer is not None else player
            piles[collector] = piles[collector] + table
            table = ()
    elif not hands[next_player]:
        # A bonus turn can run one player's hand out ahead of the other's;
        # whoever still holds cards keeps the turn.
        next_player = 1 - next_player

    return State(
        hands=tuple(hands),
        table=table,
        talon=talon,
        piles=tuple(piles),
        sweeps=tuple(sweeps),
        player=next_player,
        last_capturer=last_capturer,
    )


def deal_over(state):
    """True once every card has been taken."""
    return not (
        state.hands[0] or state.hands[1] or state.talon or state.table
    )


def score(state):
    """A pair: the points each player earned in the finished deal."""

    def points(p):
        pile = state.piles[p]
        return (
            (3 if len(pile) >= 27 else 0)
            + (2 if sum(c.endswith("S") for c in pile) >= 7 else 0)
            + sum(c.startswith("A") for c in pile)
            + (2 if "10D" in pile else 0)
            + (1 if "2S" in pile else 0)
            + state.sweeps[p]
        )

    return (points(0), points(1))

File: web/casino.js
/*
 * Cassino engine — a faithful JS port of casino.py, for the browser table.
 * Cards are strings: rank then suit. Ranks A 2 3 4 5 6 7 8 9 10 J Q K, suits S H D C.
 */

const RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"];
const VALUES = Object.fromEntries(RANKS.map((r, i) => [r, i + 1]));
const SUITS = ["S", "H", "D", "C"];

function cardValue(card) {
  return VALUES[card.slice(0, -1)];
}

function allCards() {
  const cards = [];
  for (const s of SUITS) for (const r of RANKS) cards.push(r + s);
  return cards;
}

function shuffled(cards) {
  const d = cards.slice();
  for (let i = d.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [d[i], d[j]] = [d[j], d[i]];
  }
  return d;
}

function newDeal(deck, first = 0) {
  const hands = [null, null];
  hands[first] = deck.slice(0, 3);
  hands[1 - first] = deck.slice(3, 6);
  return {
    hands,
    table: deck.slice(6, 10),
    talon: deck.slice(10),
    piles: [[], []],
    sweeps: [0, 0],
    player: first,
    lastCapturer: null,
  };
}

function* nonemptySubsets(cards) {
  const n = cards.length;
  for (let mask = 1; mask < (1 << n); mask++) {
    const subset = [];
    for (let i = 0; i < n; i++) if (mask & (1 << i)) subset.push(cards[i]);
    yield subset;
  }
}

function sumValues(cards) {
  return cards.reduce((total, c) => total + cardValue(c), 0);
}

function moveKey(hand, table) {
  return JSON.stringify([hand.slice().sort(), table.slice().sort()]);
}

function legalMoves(state) {
  const hand = state.hands[state.player];
  const table = state.table;
  const moves = [];
  const seen = new Set();

  function add(handArr, tableArr) {
    const h = handArr.slice().sort();
    const t = tableArr.slice().sort();
    const key = moveKey(h, t);
    if (!seen.has(key)) {
      seen.add(key);
      moves.push({ hand: h, table: t });
    }
  }

  for (const c of hand) add([c], []);

  const tableBySum = new Map();
  for (const t of nonemptySubsets(table)) {
    const s = sumValues(t);
    if (!tableBySum.has(s)) tableBySum.set(s, []);
    tableBySum.get(s).push(t);
  }

  for (const h of nonemptySubsets(hand)) {
    const s = sumValues(h);
    const candidates = tableBySum.get(s);
    if (candidates) for (const t of candidates) add(h, t);
  }

  return moves;
}

function findLegalMove(state, hand, table) {
  const key = moveKey(hand, table);
  return legalMoves(state).find((m) => moveKey(m.hand, m.table) === key) || null;
}

function play(state, move) {
  const found = findLegalMove(state, move.hand, move.table);
  if (!found) throw new Error("illegal move: " + JSON.stringify(move));
  move = found;

  const player = state.player;
  const other = 1 - player;

  const hands = [state.hands[0].slice(), state.hands[1].slice()];
  const handSet = new Set(move.hand);
  hands[player] = hands[player].filter((c) => !handSet.has(c));

  let table = state.table.slice();
  const tableWasEmpty = table.length === 0;
  const piles = [state.piles[0].slice(), state.piles[1].slice()];
  const sweeps = state.sweeps.slice();
  let lastCapturer = state.lastCapturer;

  const captured = move.table.length > 0;
  if (captured) {
    const tableSet = new Set(move.table);
    table = table.filter((c) => !tableSet.has(c));
    piles[player] = piles[player].concat(move.hand, move.table);
    lastCapturer = player;
    if (table.length === 0) sweeps[player] += 1;
  } else {
    table.push(move.hand[0]);
  }

  let nextPlayer =
    !captured && tableWasEmpty && hands[player].length > 0 ? player : other;

  let talon = state.talon.slice();
  if (hands[0].length === 0 && hands[1].length === 0) {
    if (talon.length > 0) {
      const lead = lastCapturer !== null ? lastCapturer : nextPlayer;
      if (lead === 0) {
        hands[0] = talon.slice(0, 3);
        hands[1] = talon.slice(3, 6);
      } else {
        hands[1] = talon.slice(0, 3);
        hands[0] = talon.slice(3, 6);
      }
      talon = talon.slice(6);
      nextPlayer = lead;
    } else if (table.length > 0) {
      const collector = lastCapturer !== null ? lastCapturer : player;
      piles[collector] = piles[collector].concat(table);
      table = [];
    }
  } else if (hands[nextPlayer].length === 0) {
    nextPlayer = 1 - nextPlayer;
  }

  return { hands, table, talon, piles, sweeps, player: nextPlayer, lastCapturer };
}

function dealOver(state) {
  return (
    state.hands[0].length === 0 &&
    state.hands[1].length === 0 &&
    state.talon.length === 0 &&
    state.table.length === 0
  );
}

function score(state) {
  function points(p) {
    const pile = state.piles[p];
    let s = 0;
    if (pile.length >= 27) s += 3;
    if (pile.filter((c) => c.endsWith("S")).length >= 7) s += 2;
    s += pile.filter((c) => c.startsWith("A")).length;
    if (pile.includes("10D")) s += 2;
    if (pile.includes("2S")) s += 1;
    s += state.sweeps[p];
    return s;
  }
  return [points(0), points(1)];
}

/** A reasonably competent, non-lookahead computer player. */
function chooseMove(state) {
  const moves = legalMoves(state);
  const cardsLeft = state.hands[0].length + state.hands[1].length + state.talon.length;
  const captures = moves.filter((m) => m.table.length > 0);

  if (captures.length > 0 && cardsLeft > 1) {
    captures.sort((a, b) => {
      if (a.table.length !== b.table.length) return b.table.length - a.table.length;
      const at = moveKey(a.hand, a.table);
      const bt = moveKey(b.hand, b.table);
      return at < bt ? -1 : at > bt ? 1 : 0;
    });
    return captures[0];
  }

  const placements = moves.filter((m) => m.table.length === 0);
  placements.sort((a, b) => (a.hand[0] < b.hand[0] ? -1 : a.hand[0] > b.hand[0] ? 1 : 0));
  return placements[0];
}
